drivers/adxl372/adxl372_spi.c

```diff
@@ -0,0 +1,123 @@
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "platform_drivers.h"
+#include "adxl372.h"
+
+/**
+ * Read from device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_spi_reg_read(struct adxl372_dev *dev,
+			     uint8_t reg_addr,
+			     uint8_t *reg_data)
+{
+	uint8_t buf[2];
+	int32_t ret;
+
+	buf[0] = ADXL372_REG_READ(reg_addr);
+	buf[1] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc, buf, ARRAY_SIZE(buf));
+	if (ret < 0)
+		return ret;
+
+	*reg_data = buf[1];
+
+	return ret;
+}
+
+/**
+ * Write to device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_spi_reg_write(struct adxl372_dev *dev,
+			      uint8_t reg_addr,
+			      uint8_t reg_data)
+{
+	uint8_t buf[2];
+
+	buf[0] = ADXL372_REG_WRITE(reg_addr);
+	buf[1] = reg_data & 0xFF;
+
+	return spi_write_and_read(dev->spi_desc, buf, ARRAY_SIZE(buf));
+}
+
+/**
+ * Multibyte read from device. A register read begins with the address
+ * and autoincrements for each aditional byte in the transfer.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @param count - Number of bytes to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_spi_reg_read_multiple(struct adxl372_dev *dev,
+				      uint8_t reg_addr,
+				      uint8_t *reg_data,
+				      uint16_t count)
+{
+	uint8_t buf[513];
+	int32_t ret;
+
+	if (count > 512) {
+		count = 512;
+		printf("warning: multiple read has been limited to 512 bytes");
```

Comment:
why not return err ?

Reply:
The maximum number of bytes you would ever want to read is 512, as this is the FIFO size. If the user will try to read more than this, the sw assumes that he wants to read the maximum number. Therefore a warning is printed instead of returning an error. However, this is just a preference.